application/src/ext/java/org/opentripplanner/ext/carpooling/routing/CarpoolStreetRouter.java

```diff
@@ -0,0 +1,162 @@
+import org.opentripplanner.street.service.StreetLimitationParametersService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Street routing service for carpooling insertion evaluation.
+ * <p>
+ * This router encapsulates all dependencies needed for A* street routing between
+ * coordinate pairs during carpool insertion optimization. It handles vertex linking,
+ * search configuration, and path selection for CAR mode routing.
+ *
+ * <h2>Routing Strategy</h2>
+ * <ul>
+ *   <li><strong>Mode:</strong> CAR mode for both origin and destination</li>
+ *   <li><strong>Algorithm:</strong> A* with Euclidean heuristic</li>
+ *   <li><strong>Dominance:</strong> Minimum weight</li>
+ *   <li><strong>Vertex Linking:</strong> Creates temporary vertices at coordinate locations</li>
+ *   <li><strong>Error Handling:</strong> Returns null on routing failure (logged as warning)</li>
+ * </ul>
+ *
+ * <h2>Design Rationale</h2>
+ * <p>
+ * This class follows OTP's established pattern for routing services (see {@code AccessEgressRouter},
+ * {@code GraphPathFinder}). Previously, routing logic was created as a lambda in
+ * {@code DefaultCarpoolingService} and passed to {@code InsertionEvaluator}, creating tight
+ * coupling. This service class provides proper encapsulation, clear ownership, and improved
+ * testability.
+ *
+ * @see InsertionEvaluator for usage in insertion evaluation
+ * @see AccessEgressRouter
+ * @see GraphPathFinder
+ */
+public class CarpoolStreetRouter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CarpoolStreetRouter.class);
+
+  private final Graph graph;
+  private final VertexLinker vertexLinker;
+  private final StreetLimitationParametersService streetLimitationParametersService;
+  private final RouteRequest request;
+
+  /**
+   * Creates a new carpool street router.
+   *
+   * @param graph the street network graph
+   * @param vertexLinker links coordinates to graph vertices
+   * @param streetLimitationParametersService provides street routing parameters (speed limits, etc.)
+   * @param request the route request containing preferences and timing
+   */
+  public CarpoolStreetRouter(
+    Graph graph,
+    VertexLinker vertexLinker,
+    StreetLimitationParametersService streetLimitationParametersService,
+    RouteRequest request
+  ) {
+    this.graph = graph;
+    this.vertexLinker = vertexLinker;
+    this.streetLimitationParametersService = streetLimitationParametersService;
+    this.request = request;
+  }
+
+  /**
+   * Routes from one location to another using A* street search.
+   * <p>
+   * Creates temporary vertices at the given coordinates, performs A* search,
+   * and returns the best path found. Returns null if routing fails.
+   *
+   * @param from origin coordinate
+   * @param to destination coordinate
+   * @return the best path found, or null if routing failed
+   */
+  public GraphPath<State, Edge, Vertex> route(GenericLocation from, GenericLocation to) {
```

Comment:
Is this actually used for routing from origin to destination, or do you use this to route between different carpooling stops?